### PR TITLE
Enable Mesh Revolving

### LINKED ISSUE #27699: Add Mesh Revolving Capability
### Motivation

MOOSE has mesh extrusion capability through [`AdvancedExtruderGenerator`](https://mooseframework.inl.gov/source/meshgenerators/AdvancedExtruderGenerator.html) and [`MeshExtruderGenerator`](https://mooseframework.inl.gov/source/meshgenerators/MeshExtruderGenerator.html) to increase mesh dimension.

Revolving is another way to increase mesh dimension. A Revolving generator that provides similar features as `AdvancedExtruderGenerator` would allow conversion from 2D-RZ mesh to 3D mesh.

### Design

Add a Revolving mesh generator that provides similar functionalities as `AdvancedExtruderGenerator`.

### Impact

Expanding MOOSE's intrinsic 3D meshing capabilities.

### PR DESCRIPTION
**Pull request by miaoyinb**
closes #27699

### CODE REVIEW

**Comment by lindsayad** [comment]
I'm very confused about why we are seeing this coverage diff in geomsearch

**Comment by miaoyinb** [comment]
> I'm very confused about why we are seeing this coverage diff in geomsearch

Yes, I am not aware of any changes here that could affect geomsearch...

File: framework/include/utils/MooseMeshUtils.h [hunk 1]
```diff
@@ -316,4 +316,57 @@ getIDFromName(const T & name)
...
+ */
+template <typename T>
+void
+idSwapParametersProcessor(const std::string & class_name,
+                          const std::string & id_name,
+                          const std::vector<std::vector<T>> & id_swaps,
+                          std::vector<std::unordered_map<T, T>> & id_swap_pairs)
+{
+  id_swap_pairs.resize(id_swaps.size());
+  for (unsigned int i = 0; i < id_swaps.size(); i++)
```
Comment (lindsayad) [inline framework/include/utils/MooseMeshUtils.h:343]:
```suggestion
  for (const auto i : index_range(id_swaps))
```

File: framework/include/utils/MooseMeshUtils.h [hunk 2]
```diff
@@ -316,4 +316,57 @@ getIDFromName(const T & name)
...
+  {
+    const std::vector<T> & swaps = id_swaps[i];
+    std::unordered_map<T, T> & swap_pairs = id_swap_pairs[i];
```
Comment (lindsayad) [inline framework/include/utils/MooseMeshUtils.h:346]:
```suggestion
    const auto & swaps = id_swaps[i];
    auto & swap_pairs = id_swap_pairs[i];
```

File: framework/src/utils/MooseMeshUtils.C [hunk 3]
```diff
@@ -484,4 +484,41 @@ makeOrderedNodeList(std::vector<std::pair<dof_id_type, dof_id_type>> & node_assm
...
+void
+extraElemIntegerSwapParametersProcessor(
+    const std::string & class_name,
+    const unsigned int num_sections,
+    const unsigned int num_integers,
+    const std::vector<std::vector<std::vector<dof_id_type>>> & elem_integers_swaps,
+    std::vector<std::unordered_map<unsigned int, unsigned int>> & elem_integers_swap_pairs)
+{
+  elem_integers_swap_pairs.resize(num_sections * num_integers);
+  for (unsigned int i = 0; i < num_integers; i++)
```
Comment (lindsayad) [inline framework/src/utils/MooseMeshUtils.C:505]:
```suggestion
  for (const auto i : make_range(num_integers))
```

File: framework/src/utils/MooseMeshUtils.C [hunk 4]
```diff
@@ -484,4 +484,41 @@ makeOrderedNodeList(std::vector<std::pair<dof_id_type, dof_id_type>> & node_assm
...
+  {
+    for (unsigned int j = 0; j < num_sections; j++)
```
Comment (lindsayad) [inline framework/src/utils/MooseMeshUtils.C:507]:
```suggestion
    for (const auto j : make_range(num_sections))
```

**Review by lindsayad — DISMISSED** [review]
reviewed the framework portion

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 5]
```diff
@@ -0,0 +1,2340 @@
...
+      "start_boundary", "The boundary ID to set on the starting boundary for partial revolving.");
+
+  params.addParam<boundary_id_type>(
+      "end_boundary", "The boundary ID to set on the ending boundary for partial revolving.");
+
+  params.addParam<bool>(
+      "clockwise", true, "Revolve clockwise around the axis or not (i.e., counterclockwise)");
+
+  params.addRequiredParam<std::vector<unsigned int>>(
+      "nums_azimuthal_intervals", "The numbers of azimuthal intervals to revolve around the axis");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:89]: is it more like the azimuthal interval discretization for each azimuthal section?

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 6]
```diff
@@ -0,0 +1,2340 @@
...
+      MooseUtils::absoluteFuzzyEqual(
+          std::accumulate(_revolving_angles.begin(), _revolving_angles.end(), 0), 360.0)
+          ? true
+          : false;
+  if (MooseUtils::absoluteFuzzyGreaterThan(
+          std::accumulate(_revolving_angles.begin(), _revolving_angles.end(), 0), 360.0))
+    paramError("revolving_angles",
+               "The sum of revolving angles should be less than or equal to 360.");
+
+  _radius_correction_factor = 1.0;
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:157]: move this to initializer list?

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 7]
```diff
@@ -0,0 +1,2340 @@
...
+std::unique_ptr<MeshBase>
+RevolveGenerator::generate()
+{
+  // Note: Inspired by AdvancedExtruderGenerator::generate()
+
+  auto mesh = buildMeshBaseObject();
+  mesh->set_mesh_dimension(_input->mesh_dimension() + 1);
+
+  // Check if the element integer names are existent in the input mesh.
+  for (unsigned int i = 0; i < _elem_integer_names_to_swap.size(); i++)
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:207]: use index_range

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 8]
```diff
@@ -0,0 +1,2340 @@
...
+    if (_input->has_elem_integer(_elem_integer_names_to_swap[i]))
+      _elem_integer_indices_to_swap.push_back(
+          _input->get_elem_integer_index(_elem_integer_names_to_swap[i]));
+    else
+      paramError("elem_integer_names_to_swap",
+                 "Element ",
+                 i + 1,
+                 " of 'elem_integer_names_to_swap' in is not a valid extra element integer of the "
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:215]:
```suggestion
                 " of 'elem_integer_names_to_swap' is not a valid extra element integer of the "
```

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 9]
```diff
@@ -0,0 +1,2340 @@
...
+                 "input mesh.");
+
+  // prepare for transferring extra element integers from original mesh to the revolved mesh.
+  const unsigned int num_extra_elem_integers = _input->n_elem_integers();
+  std::vector<std::string> id_names;
+
+  for (unsigned int i = 0; i < num_extra_elem_integers; i++)
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:222]:
```suggestion
  for (const auto i : make_range(num_extra_elem_integers))
```

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 10]
```diff
@@ -0,0 +1,2340 @@
...
+  const subdomain_id_type tri_to_pyramid_subdomain_id_shift =
+      std::max((int)max_subdomain_id, 1) + 1;
+  const subdomain_id_type tri_to_tet_subdomain_id_shift =
+      std::max((int)max_subdomain_id, 1) * 2 + 1;
+  const subdomain_id_type quad_to_prism_subdomain_id_shift = std::max((int)max_subdomain_id, 1) + 1;
+  const subdomain_id_type quad_to_pyramid_subdomain_id_shift =
+      std::max((int)max_subdomain_id, 1) * 2 + 1;
+  const subdomain_id_type quad_to_hi_pyramid_subdomain_id_shift =
+      std::max((int)max_subdomain_id, 1) * 3 + 1;
+  const subdomain_id_type edge_to_tri_subdomain_id_shift = std::max((int)max_subdomain_id, 1) + 1;
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:254]: isnt this going to overlap with quad_to_prism ?

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 11]
```diff
@@ -0,0 +1,2340 @@
...
+
+  // Get the centroid of the input mesh
+  const auto input_centroid = MooseMeshUtils::meshCentroidCalculator(*input);
+  const Point axis_centroid_cross = (input_centroid - _axis_point).cross(_axis_direction);
+
+  if (MooseUtils::absoluteFuzzyEqual(axis_centroid_cross.norm(), 0.0))
+  {
+    mooseError("The input mesh is either across the axis or overlapped with the axis!");
+  }
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:263]:
```suggestion
    mooseError("The input mesh is either across the axis or overlapped with the axis!");
```

you could use the point locator of the intial mesh to try to locate the axis in it
if the axis goes through an element  or on a node that is not on an external boundary, then error

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/RevolveGenerator.C:263]: what you have with the centroid works imo

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 12]
```diff
@@ -0,0 +1,2340 @@
...
+    }
+    // Convert the recorded subdomains
+    input->all_second_order_range(
+        input->active_subdomain_set_elements_ptr_range(converted_quad8_subdomain_ids));
+    // Restore the subdomain ID; we do not worry about repeated subdomain IDs because those QUAD9
+    // will become PYRAMID and PRISM elements with new shifts
+    for (auto elem : input->active_subdomain_set_elements_ptr_range(converted_quad8_subdomain_ids))
+    {
+      elem->subdomain_id() -= quad_to_hi_pyramid_subdomain_id_shift;
+    }
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:351]:
```suggestion
      elem->subdomain_id() -= quad_to_hi_pyramid_subdomain_id_shift;
```

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 13]
```diff
@@ -0,0 +1,2340 @@
...
+      std::accumulate(_nums_azimuthal_intervals.begin(), _nums_azimuthal_intervals.end(), 0);
+  // We know a priori how many elements we'll need
+  // In the worst case, all quad elements will become two elements per layer
+  mesh->reserve_elem(total_num_azimuthal_intervals * orig_elem * 2);
+  const dof_id_type elem_id_shift = total_num_azimuthal_intervals * orig_elem;
+
+  if (input->elements_begin() != input->elements_end() &&
+      ((*input->elements_begin())->default_order() == SECOND ||
+       (*input->elements_begin())->default_order() == THIRD))
+    order = 2;
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:380]: I changed that code in the extruder, better to do a loop
this code can fail

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 14]
```diff
@@ -0,0 +1,2340 @@
...
+  if (_preserve_volumes)
+  {
+    _radius_correction_factor = radiusCorrectionFactor(azi_array, _full_circle_revolving);
+
+    // In the meanwhile, modify the input mesh for radius correction if applicable
+    for (const auto & node : input->node_ptr_range())
+      nodeModification(*node);
+  }
+
+  mesh->comm().max(order);
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:406]: move this next to the order-determination code

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 15]
```diff
@@ -0,0 +1,2340 @@
...
+                   _unit_angles.end(),
+                   _unit_angles.begin(),
+                   [](auto & c) { return c * M_PI / 180.0; });
+  std::vector<dof_id_type> nodes_on_axis;
+
+  for (const auto & node : input->node_ptr_range())
+  {
+    // Calculate the radius and corresponding center point on the rotation axis
+    // If the radius is 0, then the node is on the axis
+    const auto radius_and_center = getRototionCenterAndRadius(*node, _axis_point, _axis_direction);
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:433]:
```suggestion
    const auto radius_and_center = getRotationCenterAndRadius(*node, _axis_point, _axis_direction);
```

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 16]
```diff
@@ -0,0 +1,2340 @@
...
+      const auto base_angle =
+          std::accumulate(_revolving_angles.begin(), _revolving_angles.begin() + e, 0.0) / 180.0 *
+          M_PI;
+
+      for (unsigned int k = 0;
+           k < order * num_layers + (e == 0 ? 1 : 0) -
+                   (e == num_rotations - 1 ? (unsigned int)_full_circle_revolving : 0);
+           ++k)
+      {
+        bool isNodeCreated(false);
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:460]:
```suggestion
        bool is_node_created(false);
```

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 17]
```diff
@@ -0,0 +1,2340 @@
...
+        {
+          // For the first layer we don't need to move
+          if (e == 0 && k == 0)
+            current_distance.zero();
+          else
+          {
+            auto layer_index = (k - (e == 0 ? 1 : 0)) + 1;
+
+            // Calculate the rotation angle in XY Plane
+
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:471]:
```suggestion
```

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 18]
```diff
@@ -0,0 +1,2340 @@
...
+      }
+    }
+  }
+
+  for (const auto & elem : input->element_ptr_range())
+  {
+    const ElemType etype = elem->type();
+
+    // revolving currently only works on coarse meshes
+    libmesh_assert(!elem->parent());
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:533]: use mooseAssert please
libmesh_assert force users to look at the code

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 19]
```diff
@@ -0,0 +1,2340 @@
...
+    for (unsigned int e = 0; e < num_rotations; e++)
+    {
+      auto num_layers = _nums_azimuthal_intervals[e];
+
+      for (unsigned int k = 0; k < num_layers; ++k)
+      {
+        std::unique_ptr<Elem> new_elem;
+        std::unique_ptr<Elem> new_elem_1;
+        bool isFlipped(false);
+        bool isFlipped_1(false);
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:548]: fix caps

add comment for what is "_1"

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 20]
```diff
@@ -0,0 +1,2340 @@
...
+                                  current_layer,
+                                  orig_nodes,
+                                  total_num_azimuthal_intervals,
+                                  side_pairs,
+                                  axis_node_case,
+                                  isFlipped);
+            }
+
+            else
+              mooseError("impossible situation");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:845]:
```suggestion
              mooseError("Degenerate element with 3 or more aligned nodes cannot be azimuthally revolved");
```

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 21]
```diff
@@ -0,0 +1,2340 @@
...
+                               new_elem,
+                               current_layer,
+                               orig_nodes,
+                               total_num_azimuthal_intervals,
+                               side_pairs,
+                               axis_node_case,
+                               isFlipped);
+            }
+            else
+              mooseError("impossible situation");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:787]: mention the element is either degenerate or unsupported, based on the node_cates

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 22]
```diff
@@ -0,0 +1,2340 @@
...
+            switch (new_elem->type())
+            {
+              case QUAD4:
+                new_elem->subdomain_id() = elem->subdomain_id();
+                break;
+              case TRI3:
+                new_elem->subdomain_id() = edge_to_tri_subdomain_id_shift + elem->subdomain_id();
+                break;
+              default:
+                mooseError("impossible situation");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:999]: these should be unreachable so they can be mooseAssert

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 23]
```diff
@@ -0,0 +1,2340 @@
...
+                                     const std::unique_ptr<MeshBase> & mesh,
+                                     std::unique_ptr<Elem> & new_elem,
+                                     const int current_layer,
+                                     const unsigned int orig_nodes,
+                                     const unsigned int total_num_azimuthal_intervals,
+                                     std::vector<std::pair<dof_id_type, dof_id_type>> & side_pairs,
+                                     bool & isFlipped)
+{
+  if (quad_elem_type != QUAD4 && quad_elem_type != QUAD9)
+    mooseError("unsupported situation");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:1521]:
```suggestion
    mooseError("Unsupported element type", quad_elem_type);
```

File: modules/reactor/src/meshgenerators/RevolveGenerator.C [hunk 24]
```diff
@@ -0,0 +1,2340 @@
...
+    std::unique_ptr<Elem> & new_elem,
+    const int current_layer,
+    const unsigned int orig_nodes,
+    const unsigned int total_num_azimuthal_intervals,
+    std::vector<std::pair<dof_id_type, dof_id_type>> & side_pairs,
+    dof_id_type & axis_node_case,
+    bool & isFlipped)
+{
+  if (tri_elem_type != TRI3 && tri_elem_type != TRI7)
+    mooseError("unsupported situation");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/RevolveGenerator.C:1585]:
```suggestion
    mooseError("Unsupported element type", tri_elem_type);
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]
saw a tokamak mesh by @cticenhour just today, this could produce a good chunk of it imo

**Comment by cticenhour** [comment]
@GiudGiud thanks for the tag! @FranklinTrevor if you have time, you might explore using this to create a single magnet to see how the workflow compares to Paramak+CUBIT.

File: modules/reactor/test/tests/meshgenerators/revolve_generator/tests [hunk 25]
```diff
@@ -0,0 +1,264 @@
...
+    []
+    [err_input_off_plane]
+      type = 'RunException'
+      input = 'revolve_1d.i'
+      cli_args = '--mesh-only err_revolve_1d.e 
+                  Mesh/rg/axis_point="0.0 0.0 1.0"'
+      expect_err = 'The input mesh is not in the same plane with the rotation axis'
+      detail = "the input 1D mesh is not in the same plane with the rotation axis."
+    []
+  []
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply modules/reactor/test/tests/meshgenerators/revolve_generator/tests:263]: I think it is banned.

[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/reactor/test/tests/meshgenerators/revolve_generator/tests:263]: it is but is there a test for it?

File: modules/reactor/src/meshgenerators/RevolveGenerator.C (hunk 10 above)
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/RevolveGenerator.C:255]: If the original element types are different (quad vs. edge), the original subdomain ids are different. In that case, it is okay to have the same shift.

**Comment by miaoyinb** [comment]
I guess the test failure is irrelevant.

**Comment by GiudGiud** [comment]
unrelated and bug fix PR is up

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
Looks good! Thanks for the contribution. I ll update the meshing project on github

**Pull request merged**